openage/convert/opus/opusenc.pyx

```diff
@@ -0,0 +1,430 @@
+
+    dbg("Wavefile: Total length: {}\tHeader length: {}\tTrailer length: {}"\
+            .format(len(inputdata), inopt['header_len'], inopt['trailer_len']))
+    if inopt['trailer_len']:
+        inputdata = inputdata[inopt['header_len']:-inopt['trailer_len']]
+    else:
+        inputdata = inputdata[inopt['header_len']:]
+
+    outdata = bytearray(1024*50)
+    cdef size_t outsize = 0  # valid bytes inside 'outdata'
+
+    cdef ogg.ogg_packet op
+    cdef ogg.ogg_page og
+    cdef ogg.ogg_stream_state os
+    cdef opus.OpusEncoder *oe
+    cdef int err
+    cdef int channels = inopt['channels']
+    cdef int wframe_len = inopt['wframe_len']
+    cdef opus.opus_int32 rate = inopt['input_rate']
+    cdef opus.opus_int32 coding_rate = 48000
+    cdef opus.opus_int32 frame_sz = 960
+    cdef opus.opus_int32 lookahead
+
+    if rate > 24000:
+        coding_rate = 48000
+    elif rate > 16000:
+        coding_rate = 24000
+    elif rate > 12000:
+        coding_rate = 16000
+    elif rate > 8000:
+        coding_rate = 12000
+    else:
+        coding_rate = 8000
+
+    frame_sz = <opus.opus_int32> (frame_sz / (48000 / coding_rate))
+
+    if coding_rate != rate:
+        dbg("Resampling necessary: input rate: {}\tcoding rate: {}".format(
+            rate, coding_rate))
+        inputdata = upsample(inputdata, inopt, coding_rate)
+        if not isinstance(inputdata, (bytes, bytearray)):
+            return "Upsampling failed: " + inputdata
+        dbg("Resampled data length: {}".format(len(inputdata)))
+
+    # initialize structs
+    if ogg.ogg_stream_init(&os, int(time.time() * 100) % (1<<31)):
+        return "Could not initialize ogg_stream."
+
+    oe = opus.opus_encoder_create(coding_rate, channels, opus.OPUS_APPLICATION_AUDIO, &err)
+    if err != opus.OPUS_OK:
+        ogg.ogg_stream_clear(&os)
+        if err == opus.OPUS_ALLOC_FAIL:
+            raise MemoryError("Could not allocate opus encoder.")
+        return "Could not allocate opus encoder."
+
+    err = opus.opus_encoder_ctl(oe, opus.OPUS_GET_LOOKAHEAD(&lookahead))
+    if err != opus.OPUS_OK:
+        free_buffers(&os, oe, NULL, NULL)
+        return "Setting encoder settings failed: {}".format(err)
+
+    inopt['pre_skip'] = lookahead / (48000 / coding_rate)
+
+    err = write_opus_header(&os, &op, inopt)
+    if err:
+        free_buffers(&os, oe, NULL, NULL)
+        if err == 1:
+            raise MemoryError("Could not create opus header.")
+        return "Could not write opus header."
+
+    while ogg.ogg_stream_flush(&os, &og):
+        outsize = write_ogg_page(outsize, outdata, &og)
+        if outsize == 0:
+            free_buffers(&os, oe, NULL, NULL)
+            raise MemoryError("Could not write opus header.")
+
+    err = write_opus_comment(&os, &op)
+    if err:
+        free_buffers(&os, oe, NULL, NULL)
+        if err == 1:
+            raise MemoryError("Could not create opus comment.")
+        return "Could not write opus comment."
```

Comment:
the typedef worked, thanks. but there is another problem. enabling the conversion from c++ exceptions to python exceptions breaks compilation. from `~/.local/lib/python3.5/site-packages/Cython/Includes/libcpp/memory.pxd`:

>     # Temporaries used for exception handling break generated code
>     unique_ptr[T] make_unique[T](...) # except + 

if I remove the hash, compilation fails:
```
/home/mcd/.repo/ap/openage/openage/convert/opus/opusenc.cpp: In function ‘PyObject* __pyx_pf_7openage_7convert_4opus_7opusenc_encode(PyObject*, PyObject*)’:
/home/mcd/.repo/ap/openage/openage/convert/opus/opusenc.cpp:2756:20: error: use of deleted function ‘std::unique_ptr<_Tp [], _Dp>& std::unique_ptr<_Tp [], _Dp>::operator=(const std::unique_ptr<_Tp [], _Dp>&) [with _Tp = char; _Dp = std::default_delete<char []>]’
       __pyx_v_spad = __pyx_t_20;
                    ^
In file included from /usr/include/c++/5/memory:81:0,
                 from /home/mcd/.repo/ap/openage/libopenage/pyinterface/../error/error.h:6,
                 from /home/mcd/.repo/ap/openage/libopenage/pyinterface/exctranslate.h:21,
                 from /home/mcd/.repo/ap/openage/libopenage/pyinterface/hacks.h:21,
                 from <command-line>:0:
/usr/include/c++/5/bits/unique_ptr.h:599:19: note: declared here
       unique_ptr& operator=(const unique_ptr&) = delete;
                   ^
```
I don't know, if someone from cython is working on that.
do we want to allow uncatched exceptions here?